src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -160,35 +175,443 @@ private DeclaredTypeAssignment GetParameterType(ParameterDeclarationSyntax synta
                 _ => null
             };
         }
 
+        private DeclaredTypeAssignment GetResourceTypeType(ResourceTypeSyntax syntax)
+            => new(GetTypeReferenceForResourceType(syntax), syntax);
+
+        private TypeSymbol GetTypeReferenceForResourceType(ResourceTypeSyntax syntax)
+        {
+            if (!features.ResourceTypedParamsAndOutputsEnabled)
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax.Span).ParamOrOutputResourceTypeUnsupported());
+            }
+
+            // The resource type of an output can be inferred.
+            var type = syntax.Type == null && GetOutputValueType(syntax) is {} inferredType ? inferredType : GetDeclaredResourceType(syntax);
+
+            if (type is ResourceType resourceType && IsExtensibilityType(resourceType))
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).UnsupportedResourceTypeParameterOrOutputType(resourceType.Name));
+            }
+
+            return type;
+        }
+
+        private bool IsExtensibilityType(ResourceType resourceType)
+        {
+            return resourceType.DeclaringNamespace.ProviderName != AzNamespaceType.BuiltInName;
+        }
+
+        private TypeSymbol? GetOutputValueType(SyntaxBase syntax) => binder.GetParent(syntax) switch
+        {
+            OutputDeclarationSyntax outputDeclaration => typeManager.GetTypeInfo(outputDeclaration.Value),
+            ParenthesizedExpressionSyntax parenthesized => GetOutputValueType(parenthesized),
+            _ => null,
+        };
+
+        private TypeSymbol ParseTypeExpression(TypeAccessSyntax syntax)
+            => binder.GetSymbolInfo(syntax) switch
+            {
+                UserDefinedTypeSymbol declaredType => TypeRefToType(syntax, declaredType),
+                DeclaredSymbol declaredSymbol => ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).ValueSymbolUsedAsType(declaredSymbol.Name)),
+                _ => ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).SymbolicNameIsNotAType(syntax.Name.IdentifierName, GetValidTypeNames())),
+            };
+
+        private IEnumerable<string> GetValidTypeNames() => LanguageConstants.DeclarationTypes.Keys
+                .Concat(binder.FileSymbol.TypeDeclarations.Select(td => td.Name))
+                .Distinct();
+
+        private TypeSymbol TypeRefToType(TypeAccessSyntax signifier, UserDefinedTypeSymbol signified)
+        {
+            var signifiedType = userDefinedTypeReferences.GetOrAdd(signified, GetUserDefinedTypeType);
+            if (signifiedType is ErrorType error)
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(signifier).ReferencedSymbolHasErrors(signified.Name));
+            }
+
+            return signifiedType;
+        }
+
+        private TypeSymbol ParseTypeExpression(ArrayTypeSyntax syntax)
```

Comment:
`Parse<xyz>` feels a bit misleading as a name for these methods - what about `Convert<xyz>` / `Generate<xyz>` / something else?

Reply:
`GetTypeFromTypeExpression`, maybe? I did want to get across that these methods assume the encountered syntax is within a type expression.